File: 0074-search-a-2d-matrix/0074-search-a-2d-matrix.java
class Solution {
    public boolean searchMatrix(int[][] matrix, int target) {
        int m = matrix.length;
        int n = matrix[0].length;

        int low = 0;
        int high = m - 1;

        while (low <= high) {
            int mid = getMidVal(low, high);
            int a = matrix[mid][0];
            int b = matrix[mid][n-1];
            if (target == a || target == b) {
                return true;
            } else if (a < target && target < b) {
                return binarySearch(matrix[mid], target);
            } else if (target < a) {
                high = mid - 1;
            } else {
                low = mid + 1;
            }
        }

        return false;
    }

    public int getMidVal(int low, int high) {
        return (int) (low + (high - low)/2);
    }

    public boolean binarySearch(int[] arr, int target) {
        int n = arr.length;
        int low = 0;
        int high = n - 1;

        while (low <= high) {
            int mid = getMidVal(low, high);

            if (arr[mid] == target) {
                return true;
            }

            if (arr[mid] < target) {
                low = mid + 1;
            } else {
                high = mid - 1;
            }
        }

        return false;
    }



}

File: 0074-search-a-2d-matrix/0074-search-a-2d-matrix.py
class Solution:

    def getTargetRow(self, matrix:[List[List[int]]], target: int) -> int:
        left, right = 0, len(matrix)-1
        while left<=right:
            mid = left + (right-left)//2
            if matrix[mid][0] == target:
                return mid
            if matrix[mid][0] > target:
                right = mid - 1
            else:
                left = mid + 1
        return right

    def searchTargetRow(self, matrix: List[List[int]], target: int, row: int):
        left = 0
        right = len(matrix[0])-1

        while left<=right:
            mid = left + (right-left)//2

            if matrix[row][mid] == target:
                return True
            
            if matrix[row][mid] > target:
                right = mid - 1
            else:
                left = mid + 1

        return False
            


    def searchMatrix(self, matrix: List[List[int]], target: int) -> bool:
        if not matrix or not matrix[0]:
            return False
    
        targetRow = self.getTargetRow(matrix, target)

        if targetRow < 0 or targetRow > len(matrix):
            return False

        if matrix[targetRow][0] <= target <= matrix[targetRow][len(matrix[0])-1]:
            return self.searchTargetRow(matrix, target, targetRow)

        return False
        